teps/0147-tekton-artifacts-phase1.md

```diff
@@ -0,0 +1,285 @@
+- [Alternatives](#alternatives)
+- [Implementation Plan (TBD)](#implementation-plan-tbd)
+  - [Test Plan](#test-plan)
+  - [Infrastructure Needed](#infrastructure-needed)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+  - [References](#references)
+<!-- /toc -->
+
+## Summary
+
+[TEP-0139](https://github.com/tektoncd/community/blob/main/teps/0139-trusted-artifacts.md) introduces the Artifacts API for the use case of artifacts (backed to a Persistent Volume) that are produced and consumed within the provenance. After more reflection and discussions, we think that it is only part of the use case of the artifacts API. It does not cover the use case of artifacts like `images`, `packages`, `source code` as discussed in [this feature request](https://github.com/tektoncd/pipeline/issues/6326). We want to be able to handle different artifact types using the same API. With `StepActions` now in play, that design needs to be updated to include `StepActions`. While the Artifact API design is still undergoing early design and brainstorming, one aspect that has general agreement is around reporting of the artifact provenance data to the `TaskRun` `Status`. We therefore think that it will be nice to split the design of Artifacts support into multiple phases and incrementally add features as needed. For now, we have identified two phases to completely capture the requirements and use cases captured in [TEP-0139](https://github.com/tektoncd/community/blob/main/teps/0139-trusted-artifacts.md) and [this feature request](https://github.com/tektoncd/pipeline/issues/6326). 
+
+This TEP proposes to define the Tekton Artifacts Provenance data structure that `Steps` can output (Phase 1). It also shows how this information is carried into the `TaskRun Status`. Furthermore, it also proposes how Tekton Chains could parse the `uri` and insert the provenance data into appropriate sections of the `SLSA` Provenance payload.
+
+## Motivation
+
+As defined in the [feature request](https://github.com/tektoncd/pipeline/issues/6326), the reporing of metadata of artifacts in Tekton has grown organically and is redundant and brittle. The structure of results make extremely challenging for a remote Task to produce artifacts of unknown length (e.g. a `ko Task` that produces images from source code does not know beforehand how many images will be produced). There is a need for users to be able to surface provenance output without being limited by the result spec.
+
+This works aims to overcome these limitations.
+
+### Goals
+#### Phase 1
+
+- Provide a mechanism for surfacing artifact provenance data to the `TaskRun` status.
+- Improve the current mechanism of collecting provenance by Tekton Chains that relies on type hinting.
+
+
+#### Phase 2
+- Provide a mechanism for making the artifact provenance data of a `TaskRun` available to downstream `TaskRuns` through variables.
```

Comment:
I still think that, with certain restrictions, we should make this part of Phase 1.

The main question I have is how to get the provenance across tasks (i.e. from the status of a Task into the disk of another Task), but I think the same question remains even with the artifact API. We could move around JSON strings, or have the entrypoint or init container fetch that via k8s API (permissions?) or have some dedicated storage.

In terms of knowing which Tasks need which provenance, the controller can know about that based on references through variables. If we introduce variables `$(tasks.<name>.[<step>].provenance)` that returns the path (or URL) to a JSON file - (or the whole JSON as a string, but that could become unwieldy in case of large provenance files), the controller can use those when setting up the DAG and when creating pods.

We would have to define what a `TaskRun` provenance is. The artifact API here would help, but there are things we can do even without it. Some ideas:
  - The provenance of the last step
  - A combination of the provenance of all the steps, with partial artifacts filtered out (assuming we can identify them)
  - We don't define a `TaskRun` provenance at all and let other tasks refer to the provenance of steps within a task directly

Reply:
If we introduce Artifacts storage as a CRD then we can enable easier referencing:

`$(tasks.<name>.<inputs/outputs>.<artifactName>)`. Over. `provenance` which could contain arbitrary inputs and outputs, a remote `Task/StepAction` will likely handle a particular type of `artifact` when consuming it. e.g. A `Task` to scan images (does not need to parse through other types of artifacts: how would it tell the difference), or a `Task` that consumes specific files from a previous `Task`. A remote `Task` probably won't know what to parse from the complete provenance file. Instead, if we enable parsing by artifact then the `Pipeline/Task` author has a reference to pass in a specific type of artifacts from a previous `Tasks`.

Reply:
Storing the artifact provenance json in the `TaskRun` CRD enables passing of individual artifact `$(tasks.<name>.<inputs/outputs>.<artifactName>[0]` into other fields (e.g. `params`, `results` etc.). Do we even need this capability? I don't think we should combine `artifacts` with `params` and `results` otherwise it gets very confusing. 

It also blows up the `TaskRun` CRD. especially if we are duplicating information between `StepState` and `TaskRun` `Status` (like we do with `StepResults` and `results`).

The downside of artifactCRD is that when it comes to referencing them, a pipeline can never first consume an artifact via the `Task API` because the information for referencing is not available. An artifact provenance must first be produced within the same pipeline before it can be consumed. But may be that's ok?

Reply:
As discussed in the Data Interfaces WG, we now have a section that handles how we can pass artifact metadata between Steps and Tasks. PTAL there.